### PR TITLE
add new edpoint to delete service account

### LINKED ISSUE #3092: New endpoint to delete SA
**User Story**
Create new endpoint to delete SA object

### PR DESCRIPTION
**Pull request by zreigz**
**What this PR does / why we need it**: new endpoint to delete service account

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3092

```release-note
 new endpoint to delete service account: DELETE /api/v1/projects/{project_id}/serviceaccounts/{serviceaccount_id}
```

### CODE REVIEW

File: api/pkg/handler/v1/serviceaccount/sa.go [hunk 1]
```diff
@@ -189,6 +210,27 @@ type updateReq struct {
 	idReq
 }
 
+// deleteReq defines HTTP request for deleteServiceAccount
+// swagger:parameters deleteServiceAccount
+type deleteReq struct {
+	idReq
+	// in: body
+	Body apiv1.ServiceAccount
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/serviceaccount/sa.go]: to delete a `SA` we only need the `ID`.

File: api/pkg/handler/v1/serviceaccount/sa.go [hunk 2]
```diff
@@ -168,6 +168,27 @@ func UpdateEndpoint(projectProvider provider.ProjectProvider, serviceAccountProv
...
+		userInfo := ctx.Value(middleware.UserInfoContextKey).(*provider.UserInfo)
+		req, ok := request.(deleteReq)
+		if !ok {
+			return nil, errors.NewBadRequest("invalid request")
+		}
+		err := req.Validate()
+		if err != nil {
+			return nil, errors.NewBadRequest(err.Error())
+		}
+
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/serviceaccount/sa.go:183]: check if the project associated with `SA` exist.

**Review by p0lyn0mial — CHANGES_REQUESTED** [review]

File: api/pkg/handler/v1/serviceaccount/sa.go (hunk 1 above)
Comment (zreigz) [reply api/pkg/handler/v1/serviceaccount/sa.go]: done

File: api/pkg/handler/v1/serviceaccount/sa.go (hunk 2 above)
Comment (zreigz) [reply api/pkg/handler/v1/serviceaccount/sa.go:183]: additionally I check if `SA` belongs to the project

File: api/pkg/handler/v1/serviceaccount/sa.go [hunk 3]
```diff
@@ -168,6 +168,49 @@ func UpdateEndpoint(projectProvider provider.ProjectProvider, serviceAccountProv
...
+		existingSA, err := serviceAccountProvider.Get(userInfo, req.ServiceAccountID)
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		project, err := projectProvider.Get(userInfo, req.ProjectID, &provider.ProjectGetOptions{})
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/serviceaccount/sa.go]: shouldn't we check if the project exist before checking/getting the `SA` (line 184) ?

File: api/pkg/handler/v1/serviceaccount/sa.go [hunk 4]
```diff
@@ -168,6 +168,49 @@ func UpdateEndpoint(projectProvider provider.ProjectProvider, serviceAccountProv
...
+		if err != nil {
+			return nil, common.KubernetesErrorToHTTPError(err)
+		}
+
+		// check if service account belongs to the project
+		existingSAList, err := serviceAccountProvider.List(userInfo, project, &provider.ServiceAccountListOptions{ServiceAccountName: existingSA.Spec.Name})
```
Comment (p0lyn0mial) [inline api/pkg/handler/v1/serviceaccount/sa.go]: we usually don't check if a resource belongs to a project like that. RBAC rules imply that a resource is part of a project.

For example, if a user can retrieve a project then a user belongs to the project. If a user can retrieve a SA then the SA must belong to the project.

However, if you really want to have a check then we could move it to the provider and examine `OwnerRef` / `labels` ?

File: api/pkg/handler/v1/serviceaccount/sa.go (hunk 3 above)
Comment (zreigz) [reply api/pkg/handler/v1/serviceaccount/sa.go]: yes, done

File: api/pkg/handler/v1/serviceaccount/sa.go (hunk 4 above)
Comment (zreigz) [reply api/pkg/handler/v1/serviceaccount/sa.go]: it's fine, I won't check

**Comment by zreigz** [comment]
@p0lyn0mial PTAL

**Pull request merged**